### ISSUE TITLE
Race condition bug in Formatter when requesting formatted changes for a sub-span in a file

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#67887]
As part of testing the changes in #67818, I have identified a race condition bug in the Formatter wherein it intermittently reports (incorrect) non-zero formatting changes when asked for formatting changes for a sub-span but reports zero formatting changes when asked for full file span.

Test code where I get wrong IDE0055 violations from lightbulb (span-based) invocation: 

![image](https://user-images.githubusercontent.com/10605811/233269868-3276527b-e6ae-4a2e-b3d5-ba48c6d35ce6.png)

Below is the local (temporary) product change in [FormattingAnalyzerHelper.cs](https://github.com/dotnet/roslyn/blob/main/src/Analyzers/Core/Analyzers/Formatting/FormattingAnalyzerHelper.cs) to handle optional FilterSpan from analysis context. For the above case, I am hitting the breakpoint at line `formattingBug = true`, indicating the formatter race condition bug I explained above:

![image](https://user-images.githubusercontent.com/10605811/233270152-717e5fd4-9ca1-4fa9-9da8-ad8257b01bf9.png)

```csharp
            var node = context.FilterSpan.HasValue ? root.FindNode(context.FilterSpan.GetValueOrDefault()) : root;
            var formattingChanges = Formatter.GetFormattedTextChanges(node, formattingProvider, options, cancellationToken);

            var formattingBug = false;
            if (formattingChanges.Count > 0 && context.FilterSpan.HasValue)
            {
                var rootChanges = Formatter.GetFormattedTextChanges(root, formattingProvider, options, cancellationToken);
                if (rootChanges.Count == 0)
                    formattingBug = true;
            }
```

**Comment by mavasani** [comment]
Turns out this is not a Formatter bug, just that I was using the wrong Formatter API for get formatted text changes.

```csharp
// Wrong API
var node = context.FilterSpan.HasValue ? root.FindNode(context.FilterSpan.GetValueOrDefault()) : root;
var formattingChanges = Formatter.GetFormattedTextChanges(node, formattingProvider, options, cancellationToken);

// Correct API
var span = context.FilterSpan.HasValue ? context.FilterSpan.GetValueOrDefault() : root.FullSpan;
var spans = SpecializedCollections.SingletonEnumerable(span);
var formattingChanges = Formatter.GetFormattedTextChanges(root, spans, formattingProvider, options, cancellationToken);
```

This was showing up as a race condition because sometimes the background analysis would compute and cache full document diagnostics before the lightbulb span based diagnostics are requested. After making the above changes, I no longer see this issue.
